crdt/sequence.py

```diff
@@ -0,0 +1,247 @@
+from functools import cmp_to_key
+from re import S
+import uuid
+from enum import Enum
+
+from crdt.gset import GSet
+from crdt.gcounter import GCounter
+
+class Sequence():
+    """
+    Sequence is a CRDT that represents an ordered set of objects and supports insertion
+    and removal at arbitrary positions in the set.
+    """
+
+    def __init__(self, id=uuid.uuid4()):
+        self.id = id
+        self.operations = GSet()
+        self.clock = GCounter(self.id)
+        self.sequence = []
+
+    def compare_operations(self, a, b):
+        """
+        Compares this Operation to another Operation to determine ordering.
+        """
+        return a.owner.compare(b.owner)
+
+    def index_transform(self, position):
+        """
+        Returns the transformed index, which is the actual index of the object in the
+        sequence, taking into account tombstones.
+        """
+        if position < 0 or position >= len(self.sequence):
+            raise IndexError("Index out of bounds")
+
+        index = -1
+        while position >= 0:
+            index += 1
+            if not self.sequence[index].tombstone:
+                position -= 1
+        return index
+
+    def append(self, object):
+        """
+        Appends an item to the end of the sequence.
+        """
+        
+        # Tick the clock
+        self.clock.add(1)
+
+        # Add the insert operation to the log and update the sequence
+        op = Operation(OpId(self.id, self.clock.get()), OperationType.INSERT, None, payload=object)
+        self.operations.add(op)
+        op.do(self.sequence)
```

Comment:
A quirk (or feature?) of this implementation is that `append` works very differently from `insert` in terms of object ordering. `insert` means "insert before this object" and `append` means "insert at the end". This means that writes from different nodes can get interleaved on appends, while for `inserts` writes have a direct relationship to another object. I think this is why Peritext/automerge defines appends as referencing the last object instead of "put this at the end", because appends are very common in document editing.

Reply:
I recommend adding the above comment to the docstring